File: src/content/1.8/code/scala/snippet10.scala
implicit def bicompBiFunctor[
    BF[_, _], FU[_], GU[_]](
    implicit BF: Bifunctor[BF],
    FU: Functor[FU], GU: Functor[GU]) = {
  // partially-applied type BiComp
  type BiCompAB[A, B] = BiComp[BF, FU, GU, A, B]
  new Bifunctor[BiCompAB] {
    override def bimap[A, B, C, D](f1: A => C)(f2: B => D): BiCompAB[A, B] => BiCompAB[C, D] = {
      case BiComp(x) =>
        BiComp(
          BF.bimap(FU.fmap(f1))(GU.fmap(f2))(x)
        )
    }
  }
}

File: src/content/1.8/code/ocaml/snippet10.ml
module BiCompBifunctor (BF : Bifunctor) (FU : Functor) (GU : Functor) =
Bifunctor_From_Bimap (struct
  type ('a, 'b) t = ('a FU.t, 'b GU.t) BF.t

  let bimap f g x = BF.bimap (FU.fmap f) (GU.fmap g) x
end)
